conjure-http/src/server.rs

```diff
@@ -0,0 +1,310 @@
+    Header(HeaderParameter),
+}
+
+/// A path parameter.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct PathParameter(());
+
+impl PathParameter {
+    /// Creates a new path parameter.
+    #[inline]
+    pub const fn new() -> PathParameter {
+        PathParameter(())
+    }
+}
+
+/// A query parameter.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct QueryParameter {
+    key: &'static str,
+}
+
+impl QueryParameter {
+    /// Creates a new query parameter.
+    #[inline]
+    pub const fn new(key: &'static str) -> QueryParameter {
+        QueryParameter { key }
+    }
+
+    /// Returns the key corresponding to this parameter in a URI's query.
+    #[inline]
+    pub fn key(&self) -> &'static str {
+        self.key
+    }
+}
+
+/// A header parameter.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct HeaderParameter {
+    header: &'static str,
+}
+
+impl HeaderParameter {
+    /// Creates a new header parameter.
+    #[inline]
+    pub const fn new(header: &'static str) -> HeaderParameter {
+        HeaderParameter { header }
+    }
+
+    /// Returns the header corresponding to this parameter in an HTTP request.
+    #[inline]
+    pub fn header(&self) -> &'static str {
+        self.header
+    }
+}
+
+/// Information about an endpoint of a resource.
+pub struct Endpoint<T, B, R>
+where
+    R: VisitResponse,
+{
+    name: &'static str,
+    method: Method,
+    path: &'static str,
+    handler: Handler<T, B, R, R::Output>,
+    parameters: &'static [Parameter],
+}
+
+impl<T, B, R> Endpoint<T, B, R>
+where
+    R: VisitResponse,
+{
+    /// Creates a new endpoint.
+    pub fn new(
+        name: &'static str,
+        method: Method,
+        path: &'static str,
+        handler: Handler<T, B, R, R::Output>,
+        parameters: &'static [Parameter],
+    ) -> Endpoint<T, B, R> {
+        Endpoint {
+            name,
+            method,
+            path,
+            handler,
+            parameters,
+        }
+    }
+
+    /// Returns the endpoint's name.
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    /// Returns the endpoint's HTTP method.
+    pub fn method(&self) -> &Method {
+        &self.method
+    }
+
+    /// Returns the endpoint's HTTP path template.
+    pub fn path(&self) -> &'static str {
+        self.path
+    }
+
+    /// Returns the endpoint's handler function pointer.
+    pub fn handler(&self) -> Handler<T, B, R, R::Output> {
+        self.handler
+    }
+
+    /// Returns the endpoint's parameters.
+    pub fn parameters(&self) -> &'static [Parameter] {
+        self.parameters
+    }
+}
+
+/// An HTTP resource.
+///
+/// The server-half of a Conjure service implements this trait.
+pub trait Resource<B, R>: Sized
+where
+    B: RequestBody,
+    R: VisitResponse,
+{
+    /// The resource's name.
+    const NAME: &'static str;
+
+    /// Returns the resource's HTTP endpoints.
+    // ideally this would be a &'static [Endpoint] once const fns become more powerful
```

Comment:
Is there an issue we can link here?

Reply:
Unclear: https://github.com/rust-lang/rust/issues/57563#issuecomment-489681952